File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.c
#include <string.h>

int distinctSubseqII(char* s) {
    long long MOD = 1000000007;

    long long dp = 1;
    long long last[26] = {0}; // Initialize all 26 elements to 0
    int n = strlen(s);

    for (int i = 0; i < n; i++) {
        int index = s[i] - 'a';

        long long previous = dp;
        // C modulo operator can yield negative values, so we add MOD before taking % MOD
        dp = (2 * dp - last[index] + MOD) % MOD;
        last[index] = previous;
    }

    return (int)((dp - 1 + MOD) % MOD);
}

File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.js
/**
 * @param {string} s
 * @return {number}
 */
var distinctSubseqII = function(s) {
    const MOD = 1000000007;

    let dp = 1;
    const last = new Array(26).fill(0);

    for (let i = 0; i < s.length; i++) {
        const index = s.charCodeAt(i) - 97; // 97 is the ASCII code for 'a'

        const previous = dp;
        // Adding MOD handles potential negative numbers safely before taking the remainder
        dp = (2 * dp - last[index] + MOD) % MOD;
        last[index] = previous;
    }

    return (dp - 1 + MOD) % MOD;
};
